### DISCUSSION TITLE
How to output '_flow_direction' in the file 'CrystalPlasticityStressUpdateBase.C' to exodus file

### DISCUSSION THREAD

**Discussion by pshen20127** [idaholab/moose#23300] [Q&A Modules: Solid mechanics] [answered]
Hello All

I want to output '_flow_direction' in the file 'CrystalPlasticityStressUpdateBase.C' to exodus file. Since it has been defined in the file 'CrystalPlasticityStressUpdateBase.C' like:
`
_flow_direction(declareProperty<std::vector<RankTwoTensor>>(_base_name + "flow_direction")),
`
So I directly add a new variable in .inp file trying to output it.

> [SchmidTensor_xx]
    type = RankTwoAux
    variable = SchmidTensor_xx
    rank_two_tensor = flow_direction
    index_j = 0
    index_i = 0
    execute_on = timestep_end
[../]

But when I run codes, it displays an error 'Material has no property named: flow_direction'. How do I output '_flow_direction'

Here is my inp file:https://github.com/pshen20127/IN625.git

Thanks
Pengfei

**Comment by GiudGiud** [comment]
Hello

The name of this property is not just flow_direction.
It's _base_name + "flow_direction"

so you need to add this base name. 

If you dont know it you can use the [Debug] option to show all material properties created

Guillaume

**Reply by sapitts** [reply]
There's not a `base_name` specified in the input file, so using `rank_two_tensor = flow_direction` ought to work

@pshen20127 I see you are using your own class in the input file. I'll see if I can replicate this error with the original `CrystalPlasticityKalidindiUpdate` class. If not, we'll probably need to see your crystal plasticity class in order to help further

**Comment by sapitts** [comment, marked as answer]
The issue here is that each slip system (in your case there are 12) has it's own Schmid tensor, so the material property `flow_direction` is a vector of Rank-2 tensors:
```
_flow_direction(declareProperty<std::vector<RankTwoTensor>>(_base_name + "flow_direction")),
```
As such, the usual `RankTwoAux` that we use for outputting tensors like the plastic deformation gradient won't work here. Same thing for the `MaterialStdVectorAux` that we use to output the slip system resistance and slip increments.

@GiudGiud do we have an auxkernel that can output a specific tensor component from a tensor within a vector of tensors?

**Reply by GiudGiud** [reply]
no I dont think we have that one

**Reply by GiudGiud** [reply]
It would not be hard to make at all

**Reply by pshen20127** [reply]
@sapitts @GiudGiud Thanks for your reply. Are all auxkernels stored in the directorties '/moose/modules/tensor_mechanics/src/auxkernels' and '/moose/framework/src/auxkernels' ? I have checked these 2 folders and don't find any works. Will you write an auxkernel that can output this type of variables? Or could you suggest other ways that can output this variable?
